client/lib/user/user.js

```diff
@@ -185,12 +171,11 @@ User.prototype.handleFetchFailure = function( error ) {
  * @param {Object} userData an object containing the user's information.
  */
 User.prototype.handleFetchSuccess = function( userData ) {
-	// Release lock from subsequent fetches
-	this.fetching = false;
 	this.clearStoreIfChanged( userData.ID );
 
-	// Store user info in `this.data` and localstorage as `wpcom_user`
-	store.set( 'wpcom_user', userData );
+	// Store user ID in local storage so that we can detect a change and clear the storage
+	store.set( 'wpcom_user_id', userData.ID );
```

Comment:
Good catch! I'm fixing that in 68f73ec, where I check the new `wpcom_user_id` key instead of the old `wpcom_user` that's gone.

One downside is that the `RedirectWhenLoggedIn` component now doesn't verify that the user with the expected ID and email has logged in, but will accept **any** login. I don't think that's a huge concern though. Implementing the check without the email being stored in local storage would be very difficult, as Calypso can't switch from logged-out to logged-in state without a full reload, and cannot retrieve the user info from the `/me` endpoint without being logged-in from the very beginning. Cc @jblz 

I tested this patch by requesting a magic login link, then logging in in second tab, and verifying that the first tab detects the login and automatically reloads and navigates to `/help`.